source/vibe/web/rest.d

```diff
@@ -762,336 +613,253 @@ private HTTPServerRequestDelegate jsonMethodHandler(alias Func, size_t ridx, T)(
+
+	static if (!is(RT == void))
+		return deserializeJson!RT(jret);
 }
 
-// Poor men's foreach (overload; MemberFunctionsTuple!(I, method))
-private mixin template RestClientMethods_OverloadImpl(Overloads...) {
-	import vibe.internal.meta.codegen : CloneFunction;
-	static assert (Overloads.length > 0);
-	//pragma(msg, "===== Body for: "~__traits(identifier, Overloads[0])~" =====");
-	//pragma(msg, genClientBody!(Overloads[0]));
-	mixin CloneFunction!(Overloads[0], genClientBody!(Overloads[0])());
-	static if (Overloads.length > 1) {
-		mixin RestClientMethods_OverloadImpl!(Overloads[1..$]);
+
+import vibe.http.client : HTTPClientRequest;
+/**
+ * Perform a request to the interface using the given parameters.
+ *
+ * Params:
+ * verb = Kind of request (See $(D HTTPMethod) enum).
+ * name = Location to request. For a request on https://github.com/rejectedsoftware/vibe.d/issues?q=author%3ASantaClaus,
+ *		it will be '/rejectedsoftware/vibe.d/issues'.
+ * hdrs = The headers to send. Some field might be overriden (such as Content-Length). However, Content-Type will NOT be overriden.
+ * query = The $(B encoded) query string. For a request on https://github.com/rejectedsoftware/vibe.d/issues?q=author%3ASantaClaus,
+ *		it will be 'author%3ASantaClaus'.
+ * body_ = The body to send, as a string. If a Content-Type is present in $(D hdrs), it will be used, otherwise it will default to
+ *		the generic type "application/json".
+ * reqReturnHdrs = A map of required return headers.
+ *				   To avoid returning unused headers, nothing is written
+ *				   to this structure unless there's an (usually empty)
+ *				   entry (= the key exists) with the same key.
+ *				   If any key present in `reqReturnHdrs` is not present
+ *				   in the response, an Exception is thrown.
+ * optReturnHdrs = A map of optional return headers.
+ *				   This behaves almost as exactly as reqReturnHdrs,
+ *				   except that non-existent key in the response will
+ *				   not cause it to throw, but rather to set this entry
+ *				   to 'null'.
+ *
+ * Returns:
+ *     The Json object returned by the request
+ */
+private Json request(URL base_url,
+	void delegate(HTTPClientRequest) request_filter, HTTPMethod verb,
+	string name, in ref InetHeaderMap hdrs, string query, string body_,
+	ref InetHeaderMap reqReturnHdrs, ref InetHeaderMap optReturnHdrs)
+{
+	import vibe.http.client : HTTPClientRequest, HTTPClientResponse, requestHTTP;
+	import vibe.http.common : HTTPStatusException, HTTPStatus, httpMethodString, httpStatusText;
+	import vibe.inet.url : Path;
+
+	URL url = base_url;
+
+	if (name.length)
+	{
+		if (url.pathString.length && url.pathString[$ - 1] == '/'
+			&& name[0] == '/')
+			url.pathString = url.pathString ~ name[1 .. $];
+		else if (url.pathString.length && url.pathString[$ - 1] == '/'
+				 || name[0] == '/')
+			url.pathString = url.pathString ~ name;
+		else
+			url.pathString = url.pathString ~ '/' ~ name;
```

Comment:
I'd really like to have a way to avoid allocations here (and in other parts of the module). At work we're using buffer + memory stomping (which was the default in D1 Tango's runtime), and can be done in D2 using [assumeSafeAppend](http://dlang.org/phobos/object.html#.assumeSafeAppend)